doc/aggregation.md

```diff
@@ -0,0 +1,164 @@
+  - Stores point/1h for one year
+  - Older points than any of mentioned age are overwriten by new incoming points
+- Each archive filled up simultaneously
+- Aggregation on the fly during writing
+- `xFilesFactor` controls if points from archive(N) should be aggregated into archive(N+1)
+- Points are selected only from one archive, with the most precision:
+  - from <= now-1d -> archive1
+  - from <= now-7d -> archive2
+  - else -> archive3
+
+## Historical graphite-clickhouse way
+
+### Storing: GraphiteMergeTree table engine in ClickHouse (CH)
+
+Completely another principle of data storage.
+
+- Retention scheme looks slightly different:  
+`retention: 0:60,1d:5m,1w:1h,1y:1d`
+  - Stores point/minute, if the age of point is at least 0sec
+  - Stores point/5min, if the age of point is at least one day
+  - Stores point/1h, if the age of point is at least one week
+  - GraphiteMergeTree doesn't drop metrics after some particular age, so after one year we would store it with the minimum possible resolution point/day
+- Retention and aggregation policies are applied only when point becomes older than X (1d,1w,1y)
+- There is no such thing as `archive`, each point is stored only once
+- No `xFilesFactor` entity: each point will be aggregated
+
+### Fetching data: before September 2019 (current `internal-aggregation = false` behavior)
+
+```sql
+SELECT Path, Time, Value, Timestamp
+FROM data WHERE ...
+```
+
+- Select all points
+- Aggregate them in *graphite-clickhouse* to the proper `archive` step
+- Pass further to *graphite-web*/*carbonapi*
+
+Problems:
+
+- A big overhead for Path (the heaviest part)
+- Overkill by network traffic, especially when CH cluster is used
+  - The CH node `query-initiator` must collect the whole data (in memory or on the disk), and only then the date will be passed further
+
+### Fetching data: after September 2019 ([#61](https://github.com/lomik/graphite-clickhouse/pull/61), [#62](https://github.com/lomik/graphite-clickhouse/pull/62), [#65](https://github.com/lomik/graphite-clickhouse/pull/61))
+
+```sql
+SELECT Path,
+  groupArray(Time),
+  groupArray(Value),
+  groupArray(Timestamp)
+FROM data WHERE ... GROUP BY Path
+```
+
+- Network consumption was decreased up to 6 times
+- But still selects all points and aggregates in *graphite-clickhouse*
+
+### Fetching data: September 2020 ([#88](https://github.com/lomik/graphite-clickhouse/pull/88))
+
+```sql
+SELECT Path,
+  arrayFilter(x->isNotNull(x),
+    anyOrNullResample($from, $until, $step)
+      (toUInt32(intDiv(Time, $step)*$step), Time)
+  ),
+  arrayFilter(x->isNotNull(x),
+    ${func}OrNullResample($from, $until, $step)
+      (Value, Time)
+  )
+FROM data WHERE ... GROUP BY Path
+```
+
+- This solution implements `archive` analog on CH side
+- The most of data is aggregated on CH shards and doesn't leave them, so `query-initiator` consumes much less memory
+- Together with *carbonapi* the `/render?maxDatePoints=x` parameter as well processed on CH side
```

Comment:
I don't really understand what is `parameter as well processed` here.

Reply:
does it sound better in 9d7f199?